watchtower/wtdb/migration4/client_db.go

```diff
@@ -0,0 +1,633 @@
+}
+
+// getMigrationStats iterates over all sessions. It counts the total number of
+// sessions as well as the total number of unmigrated sessions.
+func getMigrationStats(tx kvdb.RTx) (uint64, uint64, error) {
+	var (
+		total      uint64
+		unmigrated uint64
+	)
+
+	// Get sessions bucket.
+	mainSessionsBkt := tx.ReadBucket(cSessionBkt)
+	if mainSessionsBkt == nil {
+		return 0, 0, ErrUninitializedDB
+	}
+
+	// Iterate over each session ID in the bucket.
+	err := mainSessionsBkt.ForEach(func(sessID, _ []byte) error {
+		// Get the bucket for this particular session.
+		sessionBkt := mainSessionsBkt.NestedReadBucket(sessID)
+		if sessionBkt == nil {
+			return ErrClientSessionNotFound
+		}
+
+		total++
+
+		// Get the cSessionAckRangeIndex bucket.
+		sessionAcksBkt := sessionBkt.NestedReadBucket(cSessionAcks)
+
+		// Get the cSessionAckRangeIndex bucket.
+		sessionAckRangesBkt := sessionBkt.NestedReadBucket(
+			cSessionAckRangeIndex,
+		)
+
+		// If both buckets do not exist, then this session is empty and
+		// does not need to be migrated.
+		if sessionAckRangesBkt == nil && sessionAcksBkt == nil {
+			return nil
+		}
+
+		// If the sessionAckRangesBkt is not nil, then the session has
+		// already been migrated.
+		if sessionAckRangesBkt != nil {
+			return nil
+		}
+
+		// Else the session has not yet been migrated.
+		unmigrated++
+
+		return nil
+	})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return total, unmigrated, nil
+}
+
+// getDBChanID returns the db-assigned channel ID for the given real channel ID.
+// It returns both the uint64 and byte representation.
+func getDBChanID(chanDetailsBkt kvdb.RBucket, chanID ChannelID) (uint64,
+	[]byte, error) {
+
+	chanDetails := chanDetailsBkt.NestedReadBucket(chanID[:])
+	if chanDetails == nil {
+		return 0, nil, ErrChannelNotRegistered
+	}
+
+	idBytes := chanDetails.Get(cChanDBID)
+	if len(idBytes) == 0 {
+		return 0, nil, fmt.Errorf("no db-assigned ID found for "+
+			"channel ID %s", chanID)
+	}
+
+	id, err := readBigSize(idBytes)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	return id, idBytes, nil
+}
+
+// callback defines a type that's used by the sessionIterator.
+type callback func(bkt kvdb.RwBucket) error
+
+// sessionIterator is a helper function that iterates over the main sessions
+// bucket and performs the callback function on each individual session. If a
+// seeker is specified, it will move the cursor to the given position otherwise
+// it will start from the first item.
+func sessionIterator(tx kvdb.RwTx, seeker []byte, cb callback) ([]byte, error) {
+	// Get sessions bucket.
+	mainSessionsBkt := tx.ReadWriteBucket(cSessionBkt)
+	if mainSessionsBkt == nil {
+		return nil, ErrUninitializedDB
+	}
+
+	c := mainSessionsBkt.ReadCursor()
+	k, _ := c.First()
+
+	// Move the cursor to the specified position if seeker is non-nil.
+	if seeker != nil {
+		k, _ = c.Seek(seeker)
+	}
+
+	// Start the iteration and exit on condition.
+	for k := k; k != nil; k, _ = c.Next() {
```

Comment:
Just realized the `kvdb` cursor should have a `Current()` member so it's easier to write these type of loops.